enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,234 @@
+
+2. if the operand becomes ready, the startup-monitor will link the revision to `/etc/kubernetes/static-pods/last-known-good`, and then remove its own pod-manifest from `/etc/kubernetes`, and hence commit suicide.
+
+As long as the startup-monitor notices an operand pod-manifest of a different revision (by checking the operator manifest in `/etc/kubernetes` and the revision annotation), it will do nothing and just wait, continuing to watch the operand pod-manifest. This is important to avoid races on startup, and avoid problems on downgrade before this mechanism was introduced.
+
+### Readiness
+
+The startup-monitor watches the mirror pod for readiness, i.e. by
+
+1. watch `/var/log/kube-apiserver/termination.log` for the first start-up attempt (beware of the race of startup-monitor startup and kube-apiserver startup). Set Reason=**NeverStartedUp** when this times out.
+2. watch `/var/log/kube-apiserver/termination.log` for more than one start-up attempt. Set Reason=**CrashLooping** if more than one is found and the monitor times out.
+3. check https://localhost:6443/healthz. Set Reason=**Unhealthy** if this is red and the monitor times out.
+4. check https://localhost:6443/readyz. Set Reason=**NotReady** if this is red and the monitor times out. Reason=EtcdUnhealthy if the etcd post-start-hook never finished. In all case: message should contain the unfinished post-start-hooks.
+4. get the mirror pod from the localhost kube-apiserver
+5. check the revision annotation is the expected one. Set Reason=**NotReady** if this is red and the monitor times out.
+6. checking status.ready. Set Reason=**NotReady** if this is red and the monitor times out.
+
+The timeout for the readiness check is defined to be 5 min for the following reasons:
+
+- this timeout is for the disaster recovery case, not for the happy case. Hence, we don't have to get into the range of 60s of maxmimal API outage, i.e. around 45s as upper bound for the happy case readiness time.
+- 5 min is already a very generous duration for kube-apiserver startup. In normal situation we are in the area of O(30s).
+
+The timeout is to be specified in the operator. Hence, another operator might want different timeouts.
+
+### Retries
+
+The readiness procedure above has the risk of considering external reasons as fatal reasons for kube-apiserver readiness. E.g. etcd unstable or kubelet not updating the mirror pod. Both would led us to believe the new revision is broken.
+This is different in HA OpenShift because there is no time limit exists for a new static pod to start up. It can keep crash-looping for hours and eventually start-up to finish the roll-out.
+
+In order to mitigate this difference in error behaviour, a failing revision in SNO will be retried to roll-out after >=N minutes, N=O(10), a reasonable time for the cluster to recover from a previous attempt. In the time in-between, the fallback mechanism will move the cluster back to the old revision, until the retry is started.
```

Comment:
I hope there is a way to force a new rollout without bumping the revision counter.

Reply:
there is.